VQVAETransformerInferer not compatible with a maximum sequence length

If we set `max_seq_length` to a value smaller than the full sequence length of the encoded image, this error is thrown upon training:

`AssertionError: you are passing in a sequence length of 12585 but your absolute positional embedding has a max sequence length of 1000
`

Fixing this during training should be straightforward, if `max_seq_length` is smaller than the total sequene length we should train on a random section of the sequence of length `max_seq_length`.

I guess that we need to make [get_likelihood](https://github.com/Project-MONAI/GenerativeModels/blob/5d532373c0bb48291e1bb454ff67bd9ee2d68a73/generative/inferers/inferer.py#L528) more like the sampling function; so it loops and predicts the logits for the next token using a previous context of length `max_seq_length-1`.

--- a/generative/inferers/inferer.py
+++ b/generative/inferers/inferer.py
@@ -458,12 +458,12 @@ class VQVAETransformerInferer(Inferer):
         seq_len = latent.shape[1]
         max_seq_len = transformer_model.max_seq_len
         if max_seq_len < seq_len:
-            start = torch.randint(low=0, high=seq_len + 1 -max_seq_len, size=(1,)).item()
+            start = torch.randint(low=0, high=seq_len + 1 - max_seq_len, size=(1,)).item()
         else:
             start = 0
-        prediction = transformer_model(x=latent[:,start:start+max_seq_len], context=condition)
+        prediction = transformer_model(x=latent[:, start : start + max_seq_len], context=condition)
         if return_latent:
-            return prediction, latent[:, start:start + max_seq_len], latent_spatial_dim
+            return prediction, latent[:, start : start + max_seq_len], latent_spatial_dim
         else:
             return prediction
 
@@ -578,19 +578,19 @@ class VQVAETransformerInferer(Inferer):
         latent = latent.long()
 
         # get the first batch, up to max_seq_length, efficiently
-        logits = transformer_model(x=latent[:,:transformer_model.max_seq_len], context=condition)
+        logits = transformer_model(x=latent[:, : transformer_model.max_seq_len], context=condition)
         probs = F.softmax(logits, dim=-1)
-        probs = torch.gather(probs, 2, latent[:, :transformer_model.max_seq_len].unsqueeze(2)).squeeze(2)
+        probs = torch.gather(probs, 2, latent[:, : transformer_model.max_seq_len].unsqueeze(2)).squeeze(2)
 
         # if we have not covered the full sequence we continue with inefficient looping
         if probs.shape[1] < latent.shape[1]:
             if verbose and has_tqdm:
-                progress_bar = tqdm(range(transformer_model.max_seq_len,seq_len+1))
+                progress_bar = tqdm(range(transformer_model.max_seq_len, seq_len + 1))
             else:
-                progress_bar = iter(range(transformer_model.max_seq_len,seq_len+1))
+                progress_bar = iter(range(transformer_model.max_seq_len, seq_len + 1))
 
             for i in progress_bar:
-                idx_cond = latent[:,i+1-transformer_model.max_seq_len:i+1]
+                idx_cond = latent[:, i + 1 - transformer_model.max_seq_len : i + 1]
                 # forward the model to get the logits for the index in the sequence
                 logits = transformer_model(x=idx_cond, context=condition)
                 # pluck the logits at the final step
@@ -598,7 +598,7 @@ class VQVAETransformerInferer(Inferer):
                 # apply softmax to convert logits to (normalized) probabilities
                 p = F.softmax(logits, dim=-1)
                 # select correct values and append
-                p = torch.gather(p, 1, idx_cond[:,-1].unsqueeze(1))
+                p = torch.gather(p, 1, idx_cond[:, -1].unsqueeze(1))
                 probs = torch.cat((probs, p), dim=1)
 
         # remove starting token probability

--- a/tests/test_vqvaetransformer_inferer.py
+++ b/tests/test_vqvaetransformer_inferer.py
@@ -104,7 +104,7 @@ class TestVQVAETransformerInferer(unittest.TestCase):
         stage_1 = VQVAE(**stage_1_params)
         max_seq_len = 3
         stage_2_params_shorter = {k: v for k, v in stage_2_params.items()}
-        stage_2_params_shorter['max_seq_len'] = max_seq_len
+        stage_2_params_shorter["max_seq_len"] = max_seq_len
         stage_2 = DecoderOnlyTransformer(**stage_2_params_shorter)
         ordering = Ordering(**ordering_params)
 
@@ -234,7 +234,7 @@ class TestVQVAETransformerInferer(unittest.TestCase):
         stage_1 = VQVAE(**stage_1_params)
         max_seq_len = 3
         stage_2_params_shorter = {k: v for k, v in stage_2_params.items()}
-        stage_2_params_shorter['max_seq_len'] = max_seq_len
+        stage_2_params_shorter["max_seq_len"] = max_seq_len
         stage_2 = DecoderOnlyTransformer(**stage_2_params_shorter)
         ordering = Ordering(**ordering_params)
 
@@ -251,6 +251,7 @@ class TestVQVAETransformerInferer(unittest.TestCase):
             inputs=input, vqvae_model=stage_1, transformer_model=stage_2, ordering=ordering
         )
         self.assertEqual(likelihood.shape, latent_shape)
+
     @parameterized.expand(TEST_CASES)
     def test_get_likelihood_resampling(
         self, stage_1_params, stage_2_params, ordering_params, input_shape, logits_shape, latent_shape

